pkg/kbld/cmd/resolve.go

```diff
@@ -141,7 +141,14 @@ func (o *ResolveOptions) updateRefsInResources(nonConfigRs []ctlres.Resource,
 			}
 
 			if o.ImagesAnnotation {
-				images = append(images, img)
+				// if Metas is null then the image was already in digest form and we didn't need to resolve
+				// it, so the annotation isn't very useful
+				if len(img.Metas) > 0 {
+					// also check for duplicates before adding
+					if !Images(images).Contains(img) {
+						images = append(images, img)
+					}
+				}
```

Comment:
seems like this is more specific to NewResourceWithImages class. i think we should move deduping logic into its Bytes()

Reply:
I started doing this, but then I realized the imageStruct type is (unless I missed something) a supporting type used only by the ResourceWithImages class, so I wonder if it wouldn't make sense to keep the filtering in imageStruct instead (rather than have ResourceWithImages reach into the struct type)?

I've updated the draft in this way, but happy to move the filtering code into ResourceWithImages instead if you prefer – I'm still getting a sense of the relationships between classes.